crates/bevy_render/src/shader/shader_reflect.rs

```diff
@@ -1,300 +1,378 @@
+        };
+        let bind_groups = reflect_bind_groups(&module, shader_stage);
 
-                let mut vertex_buffer_layout = Vec::new();
-                for vertex_attribute in vertex_attributes.drain(..) {
-                    let mut instance = false;
-                    // obtain buffer name and instancing flag
-                    let current_buffer_name = {
-                        if bevy_conventions {
-                            if vertex_attribute.name == GL_VERTEX_INDEX {
-                                GL_VERTEX_INDEX.to_string()
-                            } else {
-                                instance = vertex_attribute.name.starts_with("I_");
-                                vertex_attribute.name.to_string()
-                            }
-                        } else {
-                            "DefaultVertex".to_string()
-                        }
-                    };
-
-                    // create a new buffer descriptor, per attribute!
-                    vertex_buffer_layout.push(VertexBufferLayout {
-                        attributes: vec![vertex_attribute],
-                        name: current_buffer_name.into(),
-                        step_mode: if instance {
-                            InputStepMode::Instance
-                        } else {
-                            InputStepMode::Vertex
-                        },
-                        stride: 0,
-                    });
-                }
-
-                ShaderLayout {
-                    bind_groups,
-                    vertex_buffer_layout,
-                    entry_point: entry_point_name,
-                }
-            }
-            Err(err) => panic!("Failed to reflect shader layout: {:?}.", err),
+        ShaderLayout {
+            bind_groups,
+            vertex_buffer_layout,
+            entry_point: entry_point.name.clone(),
         }
     }
 }
 
-fn reflect_bind_group(
-    descriptor_set: &ReflectDescriptorSet,
-    shader_stage: ReflectShaderStageFlags,
-) -> BindGroupDescriptor {
-    let mut bindings = Vec::new();
-    for descriptor_binding in descriptor_set.bindings.iter() {
-        let binding = reflect_binding(descriptor_binding, shader_stage);
-        bindings.push(binding);
-    }
+fn reflect_vertex_buffer_layout(
+    function: &naga::Function,
+    module: &naga::Module,
+    bevy_conventions: bool,
+) -> Vec<VertexBufferLayout> {
+    function
+        .arguments
+        .iter()
+        .filter_map(|argument| {
+            let location = match argument.binding.as_ref()? {
+                naga::Binding::Location { location, .. } => *location,
+                _ => return None,
+            };
+            let name = argument
+                .name
+                .clone()
+                .expect("expected vertex attribute to have a name");
+
+            let instance = if bevy_conventions {
+                name.starts_with("I_")
+            } else {
+                false
+            };
```

Comment:
```rust
            let step_mode = if bevy_conventions && name.starts_with("I_") {
                InputStepMode::Instance
            } else {
                InputStepMode::Vertex
            };
```
Then remove the `let step_mode = ...` on line 82.
(Just coalesce these into one `if`/`else` branch).